pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,405 @@
+    """From a density expression, extracts expressions that need to be
+    evaluated for each source and coefficients for each expression.
+
+    For eg: when this mapper is given as ``s*(s + 2) + 3`` input,
+    it returns {s**2: 1, s: 2, 1: 3}.
+    """
+    def __init__(self, source_dependent_variables):
+        self.source_dependent_variables = source_dependent_variables
+
+    def __call__(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        return super().__call__(expr)
+
+    def map_sum(self, expr):
+        stride_dicts = [self.rec(ch) for ch in expr.children]
+
+        result = defaultdict(lambda: 0)
+        for stride_dict in stride_dicts:
+            for var, stride in stride_dict.items():
+                result[var] += stride
+        return dict(result)
+
+    def map_algebraic_leaf(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        else:
+            return {1: expr}
+
+    def map_subscript(self, expr):
+        if expr in self.source_dependent_variables or \
+                expr.aggregate in self.source_dependent_variables:
+            return {expr: 1}
+        else:
+            return {1: expr}
+
+    def map_constant(self, expr):
+        return {1: expr}
+
+    def map_product(self, expr):
+        if len(expr.children) > 2:
+            left = Product(tuple(expr.children[:2]))
+            right = Product(tuple(expr.children[2:]))
+            new_prod = Product((left, right))
+            return self.rec(new_prod)
+        elif len(expr.children) == 1:
+            return self.rec(expr.children[0])
+        elif len(expr.children) == 0:
+            return {1: 1}
+        left, right = expr.children
+        d_left = self.rec(left)
+        d_right = self.rec(right)
+        d = defaultdict(lambda: 0)
+        for k_left, v_left in d_left.items():
+            for k_right, v_right in d_right.items():
+                d[k_left*k_right] += v_left*v_right
+        return dict(d)
+
+    def map_quotient(self, expr):
+        d_num = self.rec(expr.numerator)
+        d_den = self.rec(expr.denominator)
+        if len(d_den) > 1:
+            raise ValueError
+        den_k, den_v = list(d_den.items())[0]
+
+        result = {}
+        for k in d_num.keys():
+            result[d_num[k]/den_k] /= den_v
+        return result
+
+    def map_power(self, expr):
+        d_base = self.rec(expr.base)
+        d_exponent = self.rec(expr.exponent)
+        # d_exponent should look like {1: k}
+        if len(d_exponent) > 1 or 1 not in d_exponent:
+            raise RuntimeError("nonlinear expression")
+        exp = list(d_exponent.values())[0]
+        if exp == 1:
+            return d_base
+        if len(d_base) > 1:
+            raise NotImplementedError("powers are not implemented")
+        (k, v), = d_base.items()
+        return {k**exp: v**exp}
+
+    rec = __call__
+
+
+def _syzygy_module(m, generators):
```

Comment:
```suggestion
def _syzygy_module_groebner_basis_mat(m, generators):
```

Reply:
Done in 269fe81e798d6c048222baf0423cee65ed60ccee